### PR TITLE
Plugins

### LINKED ISSUE #224: Coordinate plugins and configurations update
**Is your feature request related to a problem? Please describe.**

When list of `enabled_plugins` is updated, our controller enables the plugins through a `podExec`. The `enabled_plugins` value in the server-conf configMap is updated. This means that for an existing cluster, update to any configuration (`rabbitmq.conf` or `advanced.config`) that can only be successfully loaded if an additional corresponding plugin is running, will fail to deploy.

**Describe the solution you'd like**
The most straightforward solution I can think of is to separate `enabled_plugins` in a separate configMap, and always keep the configMap up-to-date with the latest enabled plugins.That way, when people are updating `rabbitmq.conf` or `advanced.config` that requires a new plugins to be enabled, statefulSet will be restarted and the `enabled_plugins` configMaps, which is also up-to-date will be reloaded and new plugins will be enabled at node start.

The current implementation on how the controller decides to trigger a statefulSet restart, replies on the fact that the server-conf configMap is only updated with configuration changes. That's the reason I think it's easier to just separate the `enabled_plugins` to a different configMap.

**Describe alternatives you've considered**

**Additional context**
Here are the steps to reproduce the issue:

1. Have a healthy rabbitmq k8s cluster running
2. Add the following lines to `spec` replacing the username, password and uri
`kubectl edit rabbitmqclusters.rabbitmq.com`

```
rabbitmq:
   additionalPlugins:
   - rabbitmq_shovel
  
   advancedConfig: |
        [{rabbitmq_shovel,
              [ {shovels, [ {my_first_shovel,
                [ {source,
                  [ {protocol, amqp091},
                    {uris, [ "amqp://username:password@uri"]},
                  {declarations, [ {'exchange.declare',
                                     [ {exchange, <<"my_fanout">>},
                                       {type, <<"fanout">>},
                                       durable
                                     ]},
                                   {'queue.declare',
                                     [{arguments,
                                         [{<<"x-message-ttl">>, long, 60000}]}]},
                                   {'queue.bind',
                                     [ {exchange, <<"my_fanout">>},
                                       {queue,    <<>>}
                                     ]}
                                  ]},
                  {queue, <<>>},
                  {prefetch_count, 10} ]},
              {destination,
                [ {protocol, amqp091},
                  {uris, ["amqp://username:password@uri"]},
                  {declarations, [ {'exchange.declare',
                                      [ {exchange, <<"my_direct">>},
                                          {type, <<"direct">>},
                                            durable
                                          ]}
                                  ]},
                  {publish_properties, [ {delivery_mode, 2} ]},
                  {add_forward_headers, true},
                  {publish_fields, [ {exchange, <<"my_direct">>},
                                     {routing_key, <<"from_shovel">>}
                                  ]}
                ]},
              {ack_mode, on_confirm},
              {reconnect_delay, 5}
            ]}
        ]} ]} ].
```

3. The cluster will fail to start with following errors. 
` kubectl logs <pod name>`

```
Error during startup: {error,failed_to_read_advanced_configuration_file}

08:37:24.466 [error] Error during startup: {error,failed_to_read_advanced_configuration_file}
08:37:24.466 [error]
08:37:25.467 [error] Supervisor rabbit_prelaunch_sup had child prelaunch started with rabbit_prelaunch:run_prelaunch_first_phase() at undefined exit with reason failed_to_read_advanced_configuration_file in context start_error
08:37:25.467 [error] CRASH REPORT Process <0.153.0> with 0 neighbours exited with reason: {{shutdown,{failed_to_start_child,prelaunch,failed_to_read_advanced_configuration_file}},{rabbit_prelaunch_app,start,[normal,[]]}} in application_master:init/4 line 138
{"Kernel pid terminated",application_controller,"{application_start_failure,rabbitmq_prelaunch,{{shutdown,{failed_to_start_child,prelaunch,failed_to_read_advanced_configuration_file}},{rabbit_prelaunch_app,start,[normal,[]]}}}"}
Kernel pid terminated (application_controller) ({application_start_failure,rabbitmq_prelaunch,{{shutdown,{failed_to_start_child,prelaunch,failed_to_read_advanced_configuration_file}},{rabbit_prelaunch
```

4. Also, `server-conf` configmap will not contain `rabbitmq_shovel` in the `enabled_plugins` list. 
`kubectl describe configmaps rabbitmq-server-conf`

### PR DESCRIPTION
**Pull request by harshac**
This closes #224, #259

**Note to reviewers:** remember to look at the commits in this PR and consider if they can be squashed

## Summary Of Changes
- Create a new plugins builder
- Add it to the ordered list of builder alongwith the server configmap
- Add new configmap for plugins to sts
- Refactor annotations and add unit tests
- Updates the plugins before updating server conf
- Modifies controller to restart stateful set based on the builder
- Refactors controller tests to remove duplication

## Local Testing

Please ensure you run the unit, integration and system tests before approving the PR.

To run the unit and integration tests:

```
$ make unit-tests integration-tests
```

You will need to target a k8s cluster and have the operator deployed for running the system tests.

For example, for a Kubernetes context named `dev-bunny`:
```
$ kubectx dev-bunny
$ make destroy deploy-dev
# wait for operator to be deployed
$ make system-tests
```

### CODE REVIEW

**Comment by mkuratczyk** [comment]
Unfortunately this PR breaks the feature of enabling plugins without restart. Every time I add or remove a plugin from `additionalPlugins` I see the pod restarted. This should not happen if there are no configuration changes.

Steps:
1. Deploy an instance with no `additionalPlugins`
2. `kubectl edit rabbitmq ...` and add `additionalPlugins: [rabbitmq_shovel]`
3. See the pod being terminated (it shouldn't)
4. Edit again, remove `additionalPlugins`
5. See the pod being terminated (it shouldn't)

File: controllers/rabbitmqcluster_controller.go
Comment (ChunyiLyu) [inline controllers/rabbitmqcluster_controller.go:315]: small: I think it's "annotates"?

Comment (harshac) [reply controllers/rabbitmqcluster_controller.go]: Done

File: internal/resource/rabbitmq_plugins.go
Comment (ChunyiLyu) [inline internal/resource/rabbitmq_plugins.go:19]: small: may I suggest to call it "plugins" to shorten it?

Comment (harshac) [reply internal/resource/rabbitmq_plugins.go:19]: Do you mean the name of the constant?

File: internal/resource/rabbitmq_resource_builder.go
Comment (ChunyiLyu) [inline internal/resource/rabbitmq_resource_builder.go:25]: Small: I find the name of this method a bit confusing. Without the context, the name suggests that it should be the single method needed to be called to determine whether to restart the cluster or not. However, when using the method, we still need to check the operational result.

Maybe rename it to "UpdateRequiresStsRestart"? so that it's a bit more specific? Your call.

File: internal/resource/rabbitmq_plugins.go
Comment (ChunyiLyu) [inline internal/resource/rabbitmq_plugins.go:60]: Small: method name can be renamed to `RabbitmqPluginsConfigMap` so it's less redundant and consistent with other resource builder methods.

File: internal/resource/rabbitmq_plugins.go
Comment (harshac) [reply internal/resource/rabbitmq_plugins.go:19]: I will rename it to pluginsConfig to indicate that it's a configuration.

Comment (ChunyiLyu) [reply internal/resource/rabbitmq_plugins.go:19]:
```
const pluginsConfigMapName = "plugins-conf"
```
to 
```
const pluginsConfigMapName = "plugins"
```

So that the name of the configMap is shorted

File: internal/resource/rabbitmq_plugins.go
Comment (ChunyiLyu) [inline internal/resource/rabbitmq_plugins.go:45]: I see that you are deduping the list of plugins for the value for running the command. Why not do the same thing for the values in the configMap? Am I missing something? You can extract the dedup logic and use it in `pluginsValue()`

Comment (harshac) [reply internal/resource/rabbitmq_plugins.go:45]: The config map also takes in the deduped value. The asString in plugins returns deduped list as string. Maybe should name it better to make it clearer. 

I can't think of a good name for `pluginsValue`. Can you suggest something?

File: controllers/rabbitmqcluster_controller_test.go
Comment (ChunyiLyu) [inline controllers/rabbitmqcluster_controller_test.go:784]: Is this enough to make the test not flaky? I am not sure whether the problem will only happy for `v1.RoleBinding`. If this happens to `v1.Role` or any other types, we will still have flakes. And I would like the test to reliably pass in the future even if we create a new child resource.

The test here is to make sure that child resources annotations has been updated. I would suggest either to get these object to verify that updated annotations are there, or, to assert on the operator logs.

**Review by ChunyiLyu — COMMENTED** [review]
I am happy with all the functional changes. Please see my comments regarding the flaky tests that we worked on yesterday. I don't think not asserting on the events from `v1.RoleBinding` is really going to make it not flaky anymore, since this is more of a random problem that happens to the last event published. In the future, when we add a new child resource, we will likely to have this issue again. I would either assert on the operator logs, or get all the resources to see that annotations updates are reflected.

The branch now has almost 20 commits. To keep the history clear and concise, I suggest squash the commits before merging. Maybe one commit for all the feature work on plugins, a commit for `internal/metadata` refactor, and an integration tests refactor.

File: internal/resource/rabbitmq_plugins.go
Comment (ChunyiLyu) [reply internal/resource/rabbitmq_plugins.go:45]: Sorry I missed it! I think the name is fine. I need to be more careful when reviewing the code.

**Review by ChunyiLyu — APPROVED** [review]
Manually triggered charts and system tests in PR pipeline, and all tests passed. 

Lets merge it in 👍

**Pull request merged**
